ir/src/boundary_constraints.rs

```diff
@@ -80,13 +81,16 @@ impl BoundaryConstraints {
         // add the constraint to the specified boundary for the specified trace
         let col_type = symbol_table.get_type(constraint.column())?;
         let result = match col_type {
-            IdentifierType::MainTraceColumn(col_idx) => match constraint.boundary() {
-                ast::Boundary::First => self.main_first.insert(col_idx, expr),
-                ast::Boundary::Last => self.main_last.insert(col_idx, expr),
-            },
-            IdentifierType::AuxTraceColumn(col_idx) => match constraint.boundary() {
-                ast::Boundary::First => self.aux_first.insert(col_idx, expr),
-                ast::Boundary::Last => self.aux_last.insert(col_idx, expr),
+            IdentifierType::TraceColumn(column) => match column.trace_segment() {
+                0 => match constraint.boundary() {
+                    ast::Boundary::First => self.main_first.insert(column.col_idx(), expr),
+                    ast::Boundary::Last => self.main_last.insert(column.col_idx(), expr),
+                },
+                1 => match constraint.boundary() {
+                    ast::Boundary::First => self.aux_first.insert(column.col_idx(), expr),
+                    ast::Boundary::Last => self.aux_last.insert(column.col_idx(), expr),
+                },
+                _ => unimplemented!(),
```

Comment:
I think `unimplemented` makes sense here, because the only way we should get something else is if the IR gets an AST with more trace segments, which is not currently possible given the syntax of AirScript. Therefore a SemanticError doesn't really make sense. Once we support the trace segments in the language, we'll need to implement this (and refactor the boundary contraints handling in the IR)